pkg/controller/directpvmigrationprogress/directvolumemigrationprogress_controller.go

```diff
@@ -0,0 +1,368 @@
+		if transferRate != "" {
+			pvProgress.Status.LastObservedTransferRate = transferRate
+		}
+	case !containerStatus.Ready && containerStatus.LastTerminationState.Terminated != nil && containerStatus.LastTerminationState.Terminated.ExitCode != 0:
+		// pod has a failure, report last failure reason
+		pvProgress.Status.PodPhase = kapi.PodFailed
+		// TODO: report failure
+	case !containerStatus.Ready && containerStatus.LastTerminationState.Terminated != nil && containerStatus.LastTerminationState.Terminated.ExitCode == 0:
+		// succeeded dont ever requeue
+		pvProgress.Status.PodPhase = kapi.PodSucceeded
+	case pod.Status.Phase == kapi.PodSucceeded:
+		// Its possible for the succeeded pod to not have containerStatuses at all
+		pvProgress.Status.PodPhase = kapi.PodSucceeded
+	case pod.Status.Phase == kapi.PodFailed:
+		// Its possible for the succeeded pod to not have containerStatuses at all
+		pvProgress.Status.PodPhase = kapi.PodFailed
+		pvProgress.Status.LogMessage = containerStatus.State.Terminated.Message
+	}
+
+	return nil
+}
+
+func (r *ReconcileDirectVolumeMigrationProgress) Pod(cluster *migapi.MigCluster, podReference *kapi.ObjectReference) (*kapi.Pod, error) {
+	cli, err := cluster.GetClient(r)
+	if err != nil {
+		return nil, liberr.Wrap(err)
+	}
+	pod := &kapi.Pod{}
+	err = cli.Get(context.TODO(), types.NamespacedName{
+		Namespace: podReference.Namespace,
+		Name:      podReference.Name,
+	}, pod)
+	if err != nil {
+		return nil, err
+	}
+	return pod, nil
+}
+
+func (r *ReconcileDirectVolumeMigrationProgress) GetPodLogs(cluster *migapi.MigCluster, podReference *kapi.ObjectReference, tailLines *int64, previous bool) (string, error) {
+
+	config, err := cluster.BuildRestConfig(r.Client)
+	if err != nil {
+		return "", err
+	}
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return "", err
+	}
+
+	req := clientset.CoreV1().Pods(podReference.Namespace).GetLogs(podReference.Name, &kapi.PodLogOptions{
+		TailLines: tailLines,
+		Previous:  previous,
+		Container: "rsync-client",
+	})
+	readCloser, err := req.Stream()
+	if err != nil {
+		return "", err
+	}
+
+	defer readCloser.Close()
+
+	return parseLogs(readCloser)
+}
+
+func (r *ReconcileDirectVolumeMigrationProgress) GetProgressPercent(message string) string {
+	return GetLastMatch(`\d+\%`, message)
+}
+
+func (r *ReconcileDirectVolumeMigrationProgress) GetTransferRate(message string) string {
+	// Transfer Rate
+	return GetLastMatch(`\d+\.\w*\/s`, message)
+}
+
+func GetLastMatch(regex string, message string) string {
+	r := regexp.MustCompile(regex)
+	matches := r.FindAllString(message, -1)
+	if matches != nil && len(matches) > 0 {
+		return matches[len(matches)-1]
+	}
+	return ""
+}
+
+func parseLogs(reader io.Reader) (string, error) {
+	buf := new(strings.Builder)
+	_, err := io.Copy(buf, reader)
+	if err != nil {
+		return "", err
+	}
+	data := strings.Split(buf.String(), "\n")
+	logLines := []string{}
+	for _, line := range data {
```

Comment:
if we transfer GB's of small files, won't this be a very large list? are we worried about exceeding 1.5 MB limit?

Reply:
I think not https://github.com/konveyor/mig-controller/pull/734/files#diff-7e9c060a79e5cc19929659bc606b8ae646f8751e65a54f2a9de0392d46bf25b7R255 Just grabbing the last 5 lines.

I initially thought I would be grabbing 10 lines, but looks kube does not like me sending 10 line of chunk to a CR so it did not respect the \n and escaped them.

Having 5 lines got me to the point that the kubectl get is like this
```
  status:
    conditions:
    - category: Critical
      lastTransitionTime: "2020-11-09T19:30:28Z"
      message: 'The spec.podRef pvc-migrate-bmark0/directvolumemigration-rsync-transfer-pvc-0
        must reference a valid `Pod` '
      reason: NotFound
      status: "True"
      type: InvalidPod
    lastObservedProgressPercent: 96%
    lastObservedTransferRate: 40.23MB/s
    logMessage: |-
      1.41G  94%   40.22MB/s    0:00:33 (xfr#82, to-chk=18/101)202
      1.41G  95%   40.22MB/s    0:00:33 (xfr#83, to-chk=17/101)202
      1.42G  95%   40.23MB/s    0:00:33 (xfr#84, to-chk=16/101)202
      1.42G  95%   40.23MB/s    0:00:33 (xfr#85, to-chk=15/101)202
      1.43G  96%   40.23MB/s    0:00:33 (xfr#86, to-chk=14/101)202
    observedDigest: 419049d2080899448c62e94767b4fa9dac4a055e134c86ea1e740ace3a489f7a
```